apypie/action.py

```diff
@@ -10,49 +10,66 @@
 except NameError:  # Python 3 has no basestring
     basestring = str  # pylint: disable=W0622
 
+try:
+    from typing import Optional, Any, List, TYPE_CHECKING
+except ImportError:
+    TYPE_CHECKING = False
+
+if TYPE_CHECKING:
+    from apypie.api import Api
+
 
 class Action:
     """
     Apipie Action
     """
 
     def __init__(self, name, resource, api):
+        # type: (str, str, Api) -> None
         self.name = name
         self.resource = resource
         self.api = api
 
     @property
     def apidoc(self):
+        # type: () -> dict
         resource_methods = self.api.apidoc['docs']['resources'][self.resource]['methods']
         return [method for method in resource_methods if method['name'] == self.name][0]
 
     @property
     def routes(self):
+        # type: () -> List[Route]
         return [Route(route['api_url'], route['http_method'], route['short_description']) for route in self.apidoc['apis']]
 
     @property
     def params(self):
+        # type: () -> List[Param]
         return [Param(**param) for param in self.apidoc['params']]
 
     @property
     def examples(self):
+        # type: () -> List[Example]
         return [Example.parse(example) for example in self.apidoc['examples']]
 
     def call(self, params={}, headers={}, options={}, data=None, files=None):
+        # type: (dict, dict, dict, Optional[Any], Optional[dict]) -> dict
         return self.api.call(self.resource, self.name, params, headers, options, data, files)
 
     def find_route(self, params=None):
+        # type: (Optional[dict]) -> Route
         param_keys = set(self.filter_empty_params(params).keys())
         sorted_routes = sorted(self.routes, key=lambda route: [-1 * len(route.params_in_path), route.path])
         for route in sorted_routes:
             if set(route.params_in_path) <= param_keys:
                 return route
         return sorted_routes[-1]
 
-    def validate(self, values=None, data=None, files=None):
+    def validate(self, values, data=None, files=None):
```

Comment:
this is *technically* an API break, but if you would have passed `None` here, it would have failed :facepalm: